---
title: "NMT-seq_EB+ESC: strand bias"
output:
  BiocStyle::html_document: 
    toc: false
    fig_width: 10
    fig_height: 8
---

```{r load_modules, echo=FALSE, include=FALSE}
library(data.table)
library(purrr)
library(ggplot2)
library(RColorBrewer)
library(ggpubr)
```

<!-- Define settings -->
```{r define_opts, echo=FALSE}

################
## Define I/O ##
################

io <- list()
io$basedir <- "/Users/ricard/data/NMT-seq_EB+ESC"
io$metadata <- paste0(io$basedir,"/sample_metadata.txt")
io$data <- paste0(io$basedir,"/met/raw")
io$annos_dir  <- "/Users/ricard/data/NMT-seq_EB+ESC/features/filt"
io$outdir <- "/Users/ricard/data/NMT-seq_EB+ESC/met/stats/samples"

####################
## Define options ##
####################

opts <- list()

# Select lineages
opts$day_lineage <- c(
  # Day 2
  "Day2_Epiblast",
  "Day2_Primitive Streak",
  
  # Day 4/5
  "Day4_Epiblast",
  "Day4_Primitive Streak",
  "Day4_Mesoderm",
  "Day4_Endoderm",
  "Day5_Epiblast",
  "Day5_Primitive Streak",
  "Day5_Mesoderm",
  "Day5_Endoderm",
  
  # Day 6/7
  "Day6_Endoderm",
  "Day6_Mesoderm",
  "Day6_Blood",
  "Day7_Endoderm",
  "Day7_Mesoderm",
  "Day7_Blood"
)

# Select genotypes
opts$genotype <- c(
  "WT",
  "KO"
)

# Select which cells to use
opts$cells <- fread(io$metadata) %>%
  .[,day_lineage:=paste(day,lineage10x_2, sep="_")] %>%
  .[day_lineage%in%opts$day_lineage] %>%
  .[genotype%in%opts$genotype] %>%
  .[pass_metQC==T,id_met]
```

<!-- Load sample metadata -->
```{r}
metadata <- fread(io$metadata) %>% 
  .[id_met%in%opts$cells] %>%
  .[,day_lineage:=paste(day,lineage10x_2, sep="_")]
```

<!-- Load genomic contexts metadata -->
```{r load_genomiccontexts}
if (!is.null(opts$annos)) {
  anno_dt <- lapply(opts$annos, function(anno) fread(sprintf("%s/%s.bed",io$annos_dir,anno))) %>%  
    rbindlist %>% setnames(c("chr","start","end","strand","id","anno"))
}
```

<!-- Load strand bias results -->
```{r read_stats, echo=FALSE}
strand.bias <- fread("/Users/ricard/data/NMT-seq_EB+ESC/met/strand_bias/strand_bias.tsv.gz") %>%
  .[,c("id_met","discor","cov")]
```

<!-- Load cell cycle results -->
```{r}
cell.cycle <- fread("/Users/ricard/data/NMT-seq_EB+ESC/rna/cell_cycle/cell_cycle_scran.txt.gz") %>%
  # .[phase2!="unknown"] %>%
  merge(metadata, by="id_rna")
```

<!-- Merge strand bias with cell cycle assignments -->
```{r}
foo <- merge(strand.bias, cell.cycle, by="id_met")
```

<!-- Boxplots of methylation discordance between strands -->
```{r}
to.plot <- foo #%>% .[!is.na(lineage10x_2)]

p <- ggplot(to.plot, aes(genotype, y=discor)) +
  geom_boxplot(alpha=1.0, fill="#F8766D", outlier.shape = NA) +
  facet_grid(~day, scales="free_x", space = "free_x") +
  geom_jitter(alpha=0.5, color="#F8766D", size=0.75) +
  labs(x="", y="DNA methylation discordance") +
  theme_classic() +
  theme(
    axis.text.x = element_text(colour="black",size=rel(1.2), angle=45, hjust=1, vjust=1),
    strip.background = element_blank(),
    strip.text = element_text(color="black", size=rel(1.2))
  )
print(p)

# pdf(paste0(io$outdir,"/globalmet_stages.pdf"), width=9, height=5, useDingbats = F)
# print(p)
# dev.off()
```

```{r}
p <- ggplot(to.plot[phase2!="unknown"], aes(phase2, y=discor)) +
  geom_boxplot(alpha=1.0, fill="#F8766D", outlier.shape = NA) +
  # facet_wrap(~stage, scales="free_y", nrow=1) +
  facet_wrap(~genotype+day, scales="fixed") +
  geom_jitter(alpha=0.5, color="#F8766D", size=0.50) +
  labs(x="", y="DNA methylation discordance") +
  theme_classic() +
  theme(
    axis.text.x = element_text(colour="black",size=rel(1.2), angle=45, hjust=1, vjust=1),
    strip.background = element_blank(),
    strip.text = element_text(color="black", size=rel(1.2))
  )
print(p)
```

<!-- Scatterplots of methylation discordance versus coverage -->
```{r}
# to.plot[,cov:=log2(cov)]
# ggscatter(to.plot, x="cov", y="discor", add="reg.line") + 
#   facet_wrap(~stage)
```

