---
title: "Gastrulation:"
output:
  BiocStyle::html_document: 
    toc: false
    fig_width: 10
    fig_height: 8
---

```{r load_modules, echo=FALSE, include=FALSE}
library(data.table)
library(purrr)
library(ggplot2)
library(ggpubr)


theme_pb <- function() {
  theme(
    plot.title = element_text(size=rel(1.2), color="black", hjust=0.5),
    axis.text.x = element_text(size=rel(1.4), color="black", angle=50, vjust=1, hjust=1),
    axis.text.y = element_text(size=rel(1.2), color="black"),
    axis.title.y = element_text(size=rel(1.4), color="black"),
    strip.background = element_rect(fill="#F37A71"),
    strip.text = element_text(size=rel(1.5), color="black"),
    # strip.text.x = element_text(size=rel(1.45), color="black"),
    legend.position="top",
    legend.title = element_blank()
  )
}
```

<!-- Define I/O and options -->
```{r define_opts, echo=FALSE}
source("/Users/ricard/NMT-seq_EB+ESC/metacc/boxplots/load_settings.R")
```

```{r}
# sample_metadata[,.N,by=c("celltype.mapped","culture","phenotype")]
```

<!-- Parse sample metadata -->
```{r}
sample_metadata <- sample_metadata %>%
  .[,lineage:=lineage10x_2] %>%
  
  .[,culture:=ifelse(culture=="ESC_2i","EB_Day0/1/3",culture)] %>%
  .[,culture:=ifelse(culture=="ESC_Serum","EB_Day0/1/3",culture)] %>%
  .[,culture:=ifelse(culture=="EB_Day3","EB_Day0/1/3",culture)] %>%
  # .[,culture:=ifelse(culture=="EB_Day5","EB_Day5/6/7",culture)] %>%
  .[,culture:=ifelse(culture=="EB_Day6","EB_Day5/6/7",culture)] %>%
  .[,culture:=ifelse(culture=="EB_Day7","EB_Day5/6/7",culture)] %>%
  
  # Remove lineages with small number of cells
  # .[!(culture=="EB_Day67" & lineage=="Epiblast")] %>%
  # .[!(culture=="EB_Day3" & lineage=="Mesoderm")]
  .[!(culture=="EB_Day0/1/3" & lineage%in%c("Endoderm","Mesoderm","Endoderm"))] %>%
  .[!(culture=="EB_Day5/6/7" & lineage%in%c("Ectoderm","Endoderm","Blood","Epiblast"))]
```

<!-- Load methylation data -->
```{r load_data, echo=FALSE, include=FALSE}
met_dt <- lapply(names(opts$met.annos), function(n) {
  data <- fread(cmd=sprintf("zcat < %s/%s.tsv.gz",io$met.dir,n), showProgress=F, quote="") %>%
  .[V1%in%opts$met_cells]
}) %>% rbindlist %>% setnames(c("id_met","id","anno","Nmet","Ntotal","rate"))
```

<!-- Load accessibility data -->
```{r load_data, echo=FALSE, include=FALSE}
acc_dt <- lapply(names(opts$acc.annos), function(n) {
  data <- fread(cmd=sprintf("zcat < %s/%s.tsv.gz",io$acc.dir,n), showProgress=F, quote="") %>%
  .[V1%in%opts$acc_cells]
}) %>% rbindlist %>% setnames(c("id_acc","id","anno","Nmet","Ntotal","rate"))
```

<!-- Merge data with metadata -->
```{r merge}
acc_dt <- merge(acc_dt, sample_metadata, by="id_acc") %>% droplevels()
met_dt <- merge(met_dt, sample_metadata, by="id_met") %>% droplevels()
```

<!-- Load results from the differential analysis -->
```{r}
# io$diff.met <- "/Users/ricard/data/gastrulation/met/differential/feature_level"
# io$diff.acc <- "/Users/ricard/data/gastrulation/acc/differential/feature_level"
# source("/Users/ricard/gastrulation/metaccrna/differential/load_data.R")

# source("/Users/ricard/NMT-seq_EB+ESC/metacc/differential/load_data.R")
```

<!-- Subset lineage-defining sites -->
Lineage-defining elements are defined as ChIP-seq peaks that show differential activity during germ layer commitment.
At 10% FDR, this represents around 25% of Endoderm and Mesoderm putative enhancers, as defined by the ChIP-seq.
Ectoderm enhancers show very few associations with germ layer commitment, so there is no point in subsetting them.

Methylation
```{r}
# met_dt <- met_dt %>% split(.$anno) %>%
#   map2(.,names(.), function(x,y) x[id%in%diff.met[sig==T & anno==y,id]]) %>%
#   rbindlist %>% droplevels()
```

Accessibility
```{r}
# acc_dt <- acc_dt %>% split(.$anno) %>%
#   map2(.,names(.), function(x,y) x[id%in%diff.acc[sig==T & anno==y,id]]) %>%
#   rbindlist %>% droplevels()
```

```{r}
met_dt <- met_dt %>% .[,anno:=stringr::str_replace_all(anno,opts$met.annos)]
acc_dt <- acc_dt %>% .[,anno:=stringr::str_replace_all(anno,opts$acc.annos)]
```

<!-- Filter by minimum number of measurements per cell -->
```{r}
# opts$min.acc.observations <- 5
# acc_dt <- acc_dt[,N:=.N, by=c("id_acc","lineage","anno")] %>% .[N>=opts$min.acc.observations] %>% .[,N:=NULL]
# opts$min.met.observations <- 5
# met_dt <- met_dt[,N:=.N, by=c("id_met","lineage","anno")] %>% .[N>=opts$min.met.observations] %>% .[,N:=NULL]
```

```{r}
# opts$min.acc.observations <- 5
# acc_dt <- acc_dt[,N:=.N, by=c("id","lineage","anno")] %>% .[N>=opts$min.acc.observations] %>% .[,N:=NULL]
# opts$min.met.observations <- 5
# met_dt <- met_dt[,N:=.N, by=c("id","lineage","anno")] %>% .[N>=opts$min.met.observations] %>% .[,N:=NULL]
```

<!--  -->
```{r}
# acc.stats <- fread(io$acc.stats) %>% .[,c("id_acc","mean")]
# acc_dt <- acc_dt %>% merge(foo, by="id_acc") %>%
#   .[,new_rate:=mean(rate) + lm(formula=rate~mean)[["residuals"]], by=c("id","anno")]
```

<!-- Regress out global differences in mean methylation rate -->
```{r}
# foo <- fread(io$met.stats) %>% .[,c("id_met","mean")] %>% setnames("mean","global")
# met_dt <- met_dt %>% merge(foo, by="id_met") %>%
#   .[,new_rate:=mean(rate) + lm(formula=rate~global)[["residuals"]], by=c("id","anno","culture")]
```

<!-- Load genome-wide global methylation and accessibility rates -->
```{r}
met.stats <- fread(io$met.stats) %>% .[,c("id_met","mean")] %>%
  merge(sample_metadata, by="id_met") %>% .[,context:="CG"] 

acc.stats <- fread(io$acc.stats) %>% .[,c("id_acc","mean")] %>%
  merge(sample_metadata, by="id_acc") %>% .[,context:="GC"]

stats <- rbind(
  met.stats[,c("sample","mean","context")],
  acc.stats[,c("sample","mean","context")]
) %>% merge(sample_metadata,by="sample") %>%
  .[,.(mean=mean(mean)),by=c("lineage","culture","context","phenotype")] %>%
  .[,lineage_culture:=paste(lineage,culture,sep="_")] %>%
  .[,culture_phenotype:=paste(culture,phenotype,sep="_")]
```

<!-- Methylation Boxplots -->

```{r}
foo.met <- met_dt %>%
  .[,.(rate=100*(sum(Nmet)/sum(Ntotal))),by=c("id_met","lineage","anno","phenotype","culture")] %>%
  .[,lineage:=factor(lineage, levels=names(opts$lineage.colors))]

# Regress out global differences in mean methylation rate
met.stats <- fread(io$met.stats) %>% .[,c("id_met","mean")]
foo.met <- foo.met %>% merge(met.stats, by="id_met") %>%
  .[,new_rate:=rate + lm(formula=rate~mean)[["residuals"]]]

p.met <- ggplot(foo.met, aes(x=anno, y=new_rate)) +
  geom_boxplot(aes(fill=anno), outlier.shape=NA, coef=1) +
  scale_fill_manual(values=opts$enhancer.colors) +
  labs(x="", y="Methylation (%)") +
  facet_wrap(~lineage, nrow=1, scales="fixed") +
  coord_cartesian(ylim=c(0,105)) +
  theme_classic() +
  guides(color=F, fill=F) +
  theme_pb() +
  theme(
    axis.text.x = element_blank(),
    strip.background = element_rect(fill="#F37A71")
    )
print(p.met)

# pdf(sprintf("%s/boxplots_met.pdf",io$outdir), width=7, height=4)
# print(p.met)
# dev.off()
```

<!-- Acessibility boxplots -->
```{r}
foo.acc <- acc_dt %>%
  .[,.(rate=100*(sum(Nmet)/sum(Ntotal))),by=c("id_acc","lineage","anno","phenotype","culture")] %>%
  .[,lineage:=factor(lineage, levels=names(opts$lineage.colors))]

# Regress out global differences in mean accessibility rate
acc.stats <- fread(io$acc.stats) %>% .[,c("id_acc","mean")]
foo.acc <- foo.acc %>% merge(acc.stats, by="id_acc") %>%
  .[,new_rate:=rate + lm(formula=rate~mean)[["residuals"]]]

p.acc <- ggplot(foo.acc, aes(x=anno, y=new_rate)) +
  geom_boxplot(aes(fill=anno), outlier.shape=NA, coef=1) +
  scale_fill_manual(values=opts$enhancer.colors) +
  labs(x="", y="Accessibility (%)") +
  facet_wrap(~lineage, nrow=1, scales="fixed") +
  coord_cartesian(ylim=c(9,51)) +
  theme_classic() +
  guides(color=F, fill=F) +
  theme_pb() +
  theme(
    axis.text.x = element_blank(),
    strip.background = element_rect(fill="#00BFC4")
    )
print(p.acc)

# pdf(sprintf("%s/boxplots_acc.pdf",io$outdir), width=7, height=4)
# print(p.acc)
# dev.off()
```

```{r}
p <- cowplot::plot_grid(plotlist=list(p.met,p.acc), nrow = 2)
# pdf(sprintf("%s/boxplots.pdf",io$outdir), width=8, height=6)
print(p)
# dev.off()
```



<!-- Compare WT vs KO -->

Methylation
```{r}
# Regress out global differences in mean methylation rate
met.stats <- fread(io$met.stats) %>% .[,c("id_met","mean")]

foo.met2 <- foo.met %>% 
  .[,new_rate:=lm(formula=rate~mean)[["residuals"]], by=c("anno")] %>%
  .[lineage=="Epiblast"] %>%

p.met <- ggboxplot(foo.met2, x = "phenotype", y = "rate", fill="gray70", coef=1, outlier.shape=NA) +
  labs(x="", y="Methylation (%)") +
  facet_wrap(~anno, nrow=1, scales="fixed") +
  stat_compare_means(aes(label = paste0("p = ", ..p.format..)), method="t.test", comparisons=list(c("Tet_WT","Tet_KO"))) +
  theme_classic() +
  guides(color=F, fill=F) +
  theme_pb() + theme(
    axis.text.x = element_blank(),
    strip.background = element_rect(fill="#F37A71")
  )
print(p.met)

# pdf(sprintf("%s/boxplots_met_WTvsKO.pdf",io$outdir), width=9, height=4)
# print(p.met)
# dev.off()
```

Accessibility
```{r}
foo.acc2 <- foo.acc %>% .[lineage=="Epiblast"]

p.acc <- ggboxplot(foo.acc2, x = "phenotype", y = "new_rate", fill="gray70", coef=1, outlier.shape=NA) +
  labs(x="", y="Accessibility (%)") +
  facet_wrap(~anno, nrow=1, scales="fixed") +
  stat_compare_means(aes(label = paste0("p = ", ..p.format..)), method="t.test", comparisons=list(c("Tet_WT","Tet_KO"))) +
  theme_classic() +
  guides(color=F, fill=F) +
  theme_pb() + theme(
    axis.text.x = element_blank(),
    strip.background = element_rect(fill="#00BFC4")
  )
print(p.acc)

# pdf(sprintf("%s/boxplots_acc_WTvsKO.pdf",io$outdir), width=9, height=4)
# print(p.acc)
# dev.off()
```


```{r}
p <- cowplot::plot_grid(plotlist=list(p.met,p.acc), nrow = 2)

pdf(sprintf("%s/boxplots_WTvsKO.pdf",io$outdir), width=8, height=6)
print(p)
dev.off()
```

