---
title: "EB: global statistics of DNA methylation per cell"
output:
  BiocStyle::html_document: 
    toc: false
    fig_width: 10
    fig_height: 8
---

```{r load_modules, echo=FALSE, include=FALSE}
library(data.table)
library(purrr)
library(ggplot2)
```

```{r}
theme_pub <- function() {
  theme(
    axis.title.y = element_text(colour="black", size=rel(1.3)),
    axis.title.x = element_blank(),
    axis.text.x = element_blank(),
    axis.text.y = element_text(colour="black",size=rel(1.3)),
    axis.ticks = element_line(colour="black"),
    legend.position="top",
    legend.title = element_blank(),
    legend.direction = "horizontal",
    legend.key.width=unit(1.2,"line"),
    legend.key.height=unit(1.0,"line"),
    legend.text = element_text(size=15)
  )
}
```

```{r define_opts, echo=FALSE}

## Define I/O 
io <- list()
io$basedir <- "/Users/C02RF23NFVH8/data/scnmt_eb_publication"
io$in.metadata <- paste0(io$basedir, "/sample_metadata.txt")
io$in.data <- paste0(io$basedir, "/met/cpg_level")
io$annos_dir  <- paste0(io$basedir, "/features/genomic_contexts")
io$outdir  <- paste0(io$basedir, "/met/stats/samples")

## Define options
opts <- list()

# Define which lineages to use
opts$day_lineage <- c(

  # Day 2
  "Day2_Epiblast",
  "Day2_Primitive Streak",
  
  # Day 4/5
  "Day4-5_Epiblast",
  "Day4-5_Primitive Streak",
  "Day4-5_Mesoderm",
  "Day4-5_Endoderm",
  
  # Day 6/7
  "Day6-7_Endoderm",
  "Day6-7_Mesoderm",
  "Day6-7_Blood"
)

# Define which genotypes to use
opts$genotype <- c(
  "WT"
  # "KO"
)

# Define which cells to use
opts$cells <- fread(io$in.metadata) %>%
  .[,day_lineage:=paste(day2,lineage10x_2, sep="_")] %>%
  .[day_lineage%in%opts$day_lineage & genotype%in%opts$genotype] %>%
  .[pass_metQC==T,id_met]
```

<!-- Load sample metadata -->
```{r}
metadata <- fread(io$in.metadata) %>% 
  .[id_met%in%opts$cells] %>%
  .[,lineage:=lineage10x_2]
```

<!-- Load methylation data and filter genomic contexts -->
```{r load_data_and_calculate_stats, echo=FALSE}
# stats <- data.table(expand.grid(opts$cells,c(names(opts$annos),"all"))) %>% setnames(c("id_met","anno")) %>%
#   .[,c("coverage","mean"):=as.numeric(NA)]
# 
# for (cell in opts$cells) {
#   if (file.exists(sprintf("%s/%s.tsv.gz",io$in.data,cell))) {
#     # print(sprintf("Loading %s methylation...",cell))
# 
#     # Load sample methylation data
#     data <- fread(cmd=sprintf("zcat < %s/%s.tsv.gz",io$in.data,cell), sep="\t", verbose=F, showProgress=F)
# 
#     if (ncol(data)==3) {
#       data <- data %>% setnames(c("chr","pos","rate"))
#     } else {
#       data <- data %>% .[,c("chr","pos","rate")] %>% .[,rate:=rate*100]
#     }
#     
#     # Compute genome-wide statistics
#     stats[id_met==cell & anno=="all", c("coverage","mean"):=list(nrow(data),mean(data$rate))]
# 
#     # Calculate statistics per genomic context
#     data[,c("start","end") := list(pos,pos)] %>% setkey(chr,start,end)
# 
#   } else {
#     print(sprintf("Sample %s not found for methylation",cell))
#   }
# }
# 
# # stats <- stats[complete.cases(stats)]
# # fwrite(stats, file=paste0(io$outdir,"/sample_stats.txt"), sep="\t", row.names=F, col.names=T)
```

<!-- Load pre-computed estimates -->
```{r read_stats, echo=FALSE}
stats <- fread(paste0(io$outdir,"/sample_stats.txt"))
stats <- stats %>% merge(metadata, by="id_met")
```

<!-- Plot global methylation rates per cell -->
```{r}
stats_filt <- stats %>%
  .[,day2:=factor(day2,levels=c("Day2","Day4-5","Day6-7"))]# %>%
  # .[,lineage:=factor(lineage,levels=c("Epiblast","Primitive Streak","Mesoderm","Blood","Endoderm"))]

p <- ggplot(stats_filt, aes(x=day2, y=mean)) +
  geom_boxplot(alpha=0.75, fill="#F8766D", outlier.shape = NA, coef=1) +
  geom_jitter(alpha=1.0, color="#F8766D") +
  coord_cartesian(ylim=c(5,95)) +
  ylab("Global methylation (%)") +
  theme_bw() + theme_pub() +
  theme(
    strip.background = element_blank(),
    strip.text = element_blank(),
    axis.text.x = element_text(colour="black", size=rel(1.1))
  )
print(p)

# pdf(paste0(io$outdir,"/pdf/globalmet_stages.pdf"), width=6, height=6, useDingbats = F)
# print(p)
# dev.off()
```