---
title: "Gastrulation: plot differential changes using the three omics"
output: 
  BiocStyle::html_document: 
    fig_width: 12
    fig_height: 8
---

```{r echo=FALSE, include=FALSE}
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(purrr))
suppressPackageStartupMessages(library(ggplot2))
```

```{r define_opts, echo=FALSE, include=FALSE}

## Define I/O ##
io <- list()
io$basedir <- "/Users/ricard/data/gastrulation"
io$sample_metadata <- paste0(io$basedir,"/sample_metadata.txt")
io$gene_metadata <- "/Users/ricard/data/ensembl/mouse/v87/BioMart/mRNA/Mmusculus_genes_BioMart.87.txt"
io$annos_dir <- "/Users/ricard/data/gastrulation/features/filt"
io$outdir <- "/Users/ricard/gastrulation/metaccrna/differential/out"

io$diff.rna <- "/Users/ricard/data/gastrulation/rna/differential"
io$diff.met <- "/Users/ricard/data/gastrulation/met/differential/feature_level"
io$diff.acc <- "/Users/ricard/data/gastrulation/acc/differential/feature_level"

## Define options ##
opts <- list()

# window length for the overlap between genes and features
opts$gene_window <- 50000

opts$min.rna.diff <- 1
opts$min.met.diff <- 10
opts$min.acc.diff <- 10

# FDR threshold
opts$diff.type <- 2
opts$min.fdr <- 0.10

# Define genomic annotations
opts$annos <- c(
  "H3K27ac_distal_E7.5_End_intersect12" = "Endoderm enhancers",
  "H3K27ac_distal_E7.5_Mes_intersect12" = "Mesoderm enhancers",
  "H3K27ac_distal_E7.5_Ect_intersect12" = "Ectoderm enhancers",
  "prom_2000_2000" = "Promoters"
)
```

<!-- Load results from the differential analysis -->
```{r load_data, echo=FALSE, include=FALSE}
source("/Users/ricard/gastrulation/metaccrna/differential/load_data.R")
# diff.acc <- diff.acc %>% .[sig==T]
# diff.met <- diff.met %>% .[sig==T]
```


<!-- Load annotation metadata -->
```{r}
# Load genomic context metadata
feature_metadata <- lapply(names(opts$annos), function(anno) 
  fread(sprintf("%s/%s.bed",io$annos_dir,anno))[,c(1,2,3,4,5,6)]) %>%
  rbindlist %>% setnames(c("chr","start","end","strand","id","anno"))

# Load gene metadata 
gene_metadata <- fread(io$gene_metadata) %>% 
  setnames(c("ens_id","symbol"),c("id","gene")) %>% 
  .[,chr:=as.factor(sub("chr","",chr))]
```

<!-- Parse gene and feature metadata -->
```{r}
feature_metadata_filt <- feature_metadata %>% split(.$anno) %>% 
  map2(.,names(.), function(x,y) x[id %in% c(diff.met[anno==y,id],diff.acc[anno==y,id])] ) %>%
  rbindlist

gene_metadata_filt <- gene_metadata %>% .[,c("chr","start","end","gene")] %>% 
  # .[,c("start", "end") := list(start-opts$gene_window, end+opts$gene_window)] %>% 
  setkey(chr,start,end)
```

<!-- Associate the non-genic contexts with overlapping genes -->

Methylation
```{r associate, echo=FALSE}
met_list <- list()
for (ann in unique(diff.met$anno)){

  # Subset corresponding anno
  met_tmp <- diff.met[anno == ann, ]

  # Non gene-associated feature
  if (all(grepl("ENSMUSG", unique(met_tmp$id)) == FALSE)) {

    # Extract coordiantes for methylation sites and for genes
    feature_metadata_tmp <- feature_metadata_filt[anno==ann, c("chr","start","end","id")] %>%
      .[,c("start.window","end.window") := list(start - opts$gene_window, end + opts$gene_window)] %>%
      setkey(chr,start.window,end.window)

    # Do the overlap
    ov1 <- foverlaps(gene_metadata_filt, feature_metadata_filt[anno==ann, c("chr","start","end","id")] %>% setkey(chr,start,end), nomatch=0) %>%
      .[,c("gene","id")]
    ov2 <- foverlaps(gene_metadata_filt, feature_metadata_tmp, nomatch=0) %>%
      .[,.(start_dist=abs(end-i.start), end_dist=abs(start-i.end)),by=c("gene","id")] %>%
      .[,dist:=ifelse(start_dist<end_dist,start_dist,end_dist)] %>%
      # .[.[,.I[dist==min(dist)], by="id"]$V1] %>%
      .[dist<=opts$gene_window] %>%
      .[,c("gene","id")]
    ov <- rbind(ov1,ov2) %>% unique(.)

    # Merge with methylation data
    met_list[[ann]] <- merge(met_tmp, ov, by="id", allow.cartesian=T)
  }
  # Gene-associated feature
  else if (all(grepl("ENSMUSG", unique(met_tmp$id)) == TRUE)) {
    met_list[[ann]] <- merge(met_tmp, gene_metadata[,c("id","gene")], by="id")
  }
}
diff.met <- rbindlist(met_list)
```

Accessibility
```{r associate, echo=FALSE}
acc_list <- list()
for (ann in unique(diff.acc$anno)){

  # Subset corresponding anno
  acc_tmp <- diff.acc[anno == ann, ]

  # Non gene-associated feature
  if (all(grepl("ENSMUSG", unique(acc_tmp$id)) == FALSE)) {

    # Extract coordiantes for methylation sites and for genes
    feature_metadata_tmp <- feature_metadata_filt[anno==ann, c("chr","start","end","id")] %>%
      .[,c("start.window","end.window") := list(start - opts$gene_window, end + opts$gene_window)] %>%
      setkey(chr,start.window,end.window)

    # Do the overlap
    ov1 <- foverlaps(gene_metadata_filt, feature_metadata_filt[anno==ann, c("chr","start","end","id")] %>% setkey(chr,start,end), nomatch=0) %>%
      .[,c("gene","id")]
    ov2 <- foverlaps(gene_metadata_filt, feature_metadata_tmp, nomatch=0) %>%
      .[,.(start_dist=abs(end-i.start), end_dist=abs(start-i.end)),by=c("gene","id")] %>%
      .[,dist:=ifelse(start_dist<end_dist,start_dist,end_dist)] %>%
      # .[.[,.I[dist==min(dist)], by="id"]$V1] %>%
      .[dist<=opts$gene_window] %>%
      .[,c("gene","id")]
    ov <- rbind(ov1,ov2) %>% unique(.)

    # Merge with methylation data
    acc_list[[ann]] <- merge(acc_tmp, ov, by="id", allow.cartesian=T)
  }
  # Gene-associated feature
  else if (all(grepl("ENSMUSG", unique(acc_tmp$id)) == TRUE)) {
    acc_list[[ann]] <- merge(acc_tmp, gene_metadata[,c("id","gene")], by="id")
  }
}
diff.acc <- rbindlist(acc_list)
```

<!-- Merge methylation and accessibility results -->
```{r}
# diff.met %>% .[,id:=stringr::str_replace_all(id,"_500","")]
diff.met %>% .[,anno:=stringr::str_replace_all(anno,opts$annos)]
diff.acc %>% .[,anno:=stringr::str_replace_all(anno,opts$annos)]
```

<!-- Parse results -->
If a gene has multiple distal hits, just keep the most significant
```{r}
# diff.acc <- diff.acc %>% split(.$anno) %>% map(~ .[,.SD[which.min(padj_fdr)], by=c("gene")]) %>% rbindlist
# diff.met <- diff.met %>% split(.$anno) %>% map(~ .[,.SD[which.min(padj_fdr)], by=c("gene")]) %>% rbindlist
```

```{r}
diff.metacc <- rbind(
  diff.met[,type:="met"], 
  diff.acc[,type:="acc"]
# ) %>% dcast(gene+anno~type, value.var=c("padj_fdr","diff","sig")) %>% .[complete.cases(.)]
) %>% dcast(id+gene+lineage+anno~type, value.var=c("diff","sig")) %>% .[complete.cases(.)]
```

<!-- Merge with RNA expression results -->
```{r}
diff.rna_filt <- diff.rna[,c("symbol","lineage","logFC","sig")] %>%
  setnames(c("symbol","logFC","sig"),c("gene","rna_diff","rna_sig"))

diff.metaccrna <- diff.metacc %>% merge(diff.rna_filt, by=c("gene","lineage"))
```

<!-- Filter -->
```{r}
diff.metaccrna.filt <- diff.metaccrna #[rna_sig==T]
# diff.metaccrna.filt <- diff.metacc
```

<!-- Scatterplot of differential methylation versus differential accessibiliy -->
```{r}
foo <- diff.metaccrna.filt %>% copy %>% .[,sig:=as.factor((rna_sig) & (sig_met|sig_acc))]

# foo <- diff.metaccrna.filt %>% copy %>% .[,sig:=as.factor(sig_met|sig_acc)]
```

```{r}
# opts$ymin <- -75
# opts$ymax <- 75
# 
# opts$xmin <- -75
# opts$xmax <- 75


opts$ymin <- -45
opts$ymax <- 45

opts$xmin <- -60
opts$xmax <- 60
```

```{r}
opts$colors <- c(Ectoderm="steelblue", Endoderm="#43CD80", Mesoderm="violetred")
```


<!-- Enhancers -->
```{r}
# i <- ""

for (i in unique(foo$lineage)) {
  foo_filt <- foo[lineage==i]
  
  p <- ggplot(foo_filt, aes(x=diff_met, y=diff_acc)) +
    geom_segment(aes(x=opts$xmin, xend=opts$xmax, y=0, yend=0), size=0.25, color="orange") +
    geom_segment(aes(x=0, xend=0, y=opts$ymin, yend=opts$ymax), size=0.25, color="orange") +

    # geom_point(size=0.5, data=foo_filt, color="black") +
    geom_point(size=0.8, data=foo_filt[sig==T]) +
    geom_point(size=0.5, data=foo_filt[sig==F], alpha=0.2, color="grey") +
    
    ggrepel::geom_text_repel(data=foo_filt[sig==T], aes(x=diff_met, y=diff_acc, label=gene), size=5, color=opts$colors[i]) +
    coord_cartesian(xlim=c(opts$xmin,opts$xmax), ylim=c(opts$ymin,opts$ymax)) +
    # stat_smooth(method="lm", data=foo) +
    labs(x="Differential methylation rate", y="Differential accessibility rate", title=i) +
    theme(
      axis.text.x = element_text(size=rel(1.1), color='black'),
      axis.text.y = element_text(size=rel(1.1), color='black'),
      axis.title.x = element_text(size=rel(1.2), color='black'),
      axis.title.y = element_text(size=rel(1.2), color='black'),
      legend.position = "none",
      panel.background = element_blank()
    )
  print(p)
  
  # pdf(sprintf("%s/diffmetaccrna_%s_enhancers.pdf",io$outdir,opts$lineage), width=5, height=4, useDingbats = F)
  # print(p)
  # dev.off()
}
```

<!-- Promoters -->
```{r}
# foo_filt <- foo[anno=="Promoters"]
# 
# p <- ggplot(foo_filt, aes(x=diff_met, y=diff_acc)) +
#   geom_segment(aes(x=opts$xmin, xend=opts$xmax, y=0, yend=0), size=0.25, color="orange") +
#   geom_segment(aes(x=0, xend=0, y=opts$ymin, yend=opts$ymax), size=0.25, color="orange") +
#   # geom_point(size=1.75, data=foo_filt[sig==T]) +
#   # geom_point(size=0.5, data=foo_filt[sig==F], alpha=0.1) +
#   
#   ggrastr::geom_point_rast(size=1.75, data=foo_filt[sig==T]) +
#   ggrastr::geom_point_rast(size=0.5, data=foo_filt[sig==F], alpha=0.1) +
#   
#   ggrepel::geom_text_repel(data=foo_filt[sig==T], aes(x=diff_met, y=diff_acc, label=gene), size=5, color=color) +
#   coord_cartesian(xlim=c(opts$xmin,opts$xmax), ylim=c(opts$ymin,opts$ymax)) +
#   labs(x="Differential methylation rate", y="Differential accessibility rate") +
#   theme(
#     axis.text.x = element_text(size=rel(1.1), color='black'),
#     axis.text.y = element_text(size=rel(1.1), color='black'),
#     axis.title.x = element_text(size=rel(1.2), color='black'),
#     axis.title.y = element_text(size=rel(1.2), color='black'),
#     legend.position = "none",
#     panel.background = element_blank()
#   )
# print(p)
# 
# # pdf(sprintf("%s/diffmetaccrna_%s_promoters.pdf",io$outdir,opts$lineage), width=5, height=4, useDingbats = F)
# # print(p)
# # dev.off()
```


TEST
```{r}
# i <- ""

for (i in unique(foo$lineage)) {
  foo_filt <- foo[lineage==i] %>% 
    .[,rna_diff:=abs(rna_diff)]  %>%
    .[,sig:=as.factor((rna_sig) & (sig_met))] %>%
    .[is.na(rna_diff),rna_diff:=0] %>%
    .[,diff_met:=-diff_met] %>%
    .[,diff_acc:=-diff_acc]

  
  p1 <- ggplot(foo_filt, aes(x=diff_met, y=rna_diff)) +
    geom_segment(aes(x=0, xend=0, y=0, yend=10), size=0.25, color="orange") +

    # geom_point(size=0.5, data=foo_filt, color="black") +
    geom_point(size=1.2, data=foo_filt[sig==T]) +
    geom_point(size=0.6, data=foo_filt[sig==F], alpha=0.75, color="grey") +
    
    ggrepel::geom_text_repel(data=foo_filt[sig==T], aes(x=diff_met, y=rna_diff, label=gene), size=5, color=opts$colors[i]) +
    coord_cartesian(xlim=c(-50,50)) +
    # stat_smooth(method="lm", data=foo) +
    labs(x="Differential methylation rate", y="Differential RNA expression", title=i) +
    theme_bw() +
    theme(
      plot.title = element_blank(),
      axis.text.x = element_text(size=rel(1.2), color='black'),
      axis.text.y = element_text(size=rel(1.2), color='black'),
      axis.title.x = element_text(size=rel(1.2), color='black'),
      axis.title.y = element_text(size=rel(1.2), color='black'),
      legend.position = "none"
    )
  # print(p1)
  
  p2 <- ggplot(foo_filt, aes(x=diff_met)) +
      geom_density(aes(y=..scaled..), alpha=0.6, fill="grey70") +
    labs(x="", y="Density", title="") +
    coord_cartesian(xlim=c(-50,50)) +
    theme_classic() +
    theme(
      axis.line = element_line(size=rel(0.7)),
      axis.text.x = element_blank(),
      axis.text.y = element_text(size=rel(0.8), color='black')
    )
  # print(p2)
  
  p <- cowplot::plot_grid(plotlist=list(p2,p1), nrow=2, rel_heights = c(1/4,3/4))
  
  pdf(sprintf("%s/test_%s_enhancers2.pdf",io$outdir,i), width=6, height=5, useDingbats = F)
  print(p)
  dev.off()
}
```

